### PR TITLE
Exercise1 Code Merge

### PR DESCRIPTION
**Pull request by Alex-Caesar**

### CODE REVIEW

File: exercise1/terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+      source  = "hashicorp/azurerm"
+      version = "3.99.0"
+    }
+  }
+}
+
+provider "azurerm" {
+  features {
+    resource_group {
+      prevent_deletion_if_contains_resources = true
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf:17]: This is really helpful. I did not know this was available.

File: exercise1/terraform/main.tf [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+
+#_______________________ private endpoint network resources _______________________________________________
+resource "azurerm_subnet" "ex1_subnet_pe" {
+  name                 = "${var.rg_name}_subnet_pe"
+  resource_group_name  = azurerm_resource_group.ex1.name
+  virtual_network_name = azurerm_virtual_network.ex1_vnet.name
+  address_prefixes     = ["10.0.0.0/24"]
+}
+
+resource "azurerm_network_security_group" "ex1_sql_netsecg" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: You should add more explicit rules allowing only required ports. In this case, only sql is needed.

File: exercise1/terraform/main.tf [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+  location            = azurerm_resource_group.ex1.location
+  resource_group_name = azurerm_resource_group.ex1.name
+}
+
+resource "azurerm_subnet_network_security_group_association" "ex1_secg_asso_pe" {
+  subnet_id                 = azurerm_subnet.ex1_subnet_pe.id
+  network_security_group_id = azurerm_network_security_group.ex1_sql_netsecg.id
+}
+
+resource "azurerm_private_dns_zone" "ex1_priv_dns_zone" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: DNS zones for Private Endpoints (PEs) should match the azure documentation explicitly:
https://learn.microsoft.com/en-us/azure/private-link/private-endpoint-dns#databases

File: exercise1/terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+  name                = "${var.rg_name}.privatelink.database.windows.net"
+  resource_group_name = azurerm_resource_group.ex1.name
+}
+
+resource "azurerm_private_dns_zone_virtual_network_link" "ex1_priv_dns_z_net_link" {
+  name                  = "${var.rg_name}_priv_dns_z_net_link"
+  resource_group_name   = azurerm_resource_group.ex1.name
+  registration_enabled  = true
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: This is only required when you would like to auto register VMs to this DNS zone. This is not applicable in this case.

File: exercise1/terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+  private_service_connection {
+    name                           = "${var.rg_name}_redis_private_serv_conn"
+    private_connection_resource_id = azurerm_redis_cache.ex1_vm_redis.id
+    subresource_names              = ["redisCache"]
+    is_manual_connection           = false
+  }
+
+  private_dns_zone_group {
+    name                 = "${var.rg_name}_private_dns_zg_redis"
+    private_dns_zone_ids = [azurerm_private_dns_zone.ex1_priv_dns_zone.id]
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: You need to create a private dns zone per resource type and associate that, in this case redis.
https://learn.microsoft.com/en-us/azure/private-link/private-endpoint-dns#databases

File: exercise1/terraform/main.tf [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+  private_dns_zone_group {
+    name                 = "${var.rg_name}_private_dns_zg_sql"
+    private_dns_zone_ids = [azurerm_private_dns_zone.ex1_priv_dns_zone.id]
+  }
+}
+
+# ********************** Database associated resources **************************************************
+
+#_______________________ SQL Database resources _______________________________________________
+resource "azurerm_storage_account" "ex1_store_acc" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: A Storage account was not part of the requirements. What is your reasoning for creating one?

File: exercise1/terraform/main.tf (hunk 5 above)
Comment (jb-coalfire) [reply exercise1/terraform/main.tf]: You will need to create the vnet link for the redis zone as well

File: exercise1/terraform/main.tf [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+  location            = azurerm_resource_group.ex1.location
+
+  ip_configuration {
+    name                          = local.vm_nic_ip_name
+    subnet_id                     = azurerm_subnet.ex1_subnet_vm.id
+    private_ip_address_allocation = "Dynamic"
+  }
+}
+
+resource "azurerm_network_interface_security_group_association" "ex1_vm_netsecg_nic_asso" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: you do not need to associate the nic to the NSG since you already associated the subnet where this NIC resides. So this is redundant. Good practice so associate NSGs at the subnet level.

File: exercise1/terraform/main.tf [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+
+resource "azurerm_network_security_rule" "ex1_netsec_r_vm_443" {
+  name                        = "${var.rg_name}_netsec_r_443"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "443"
+  destination_port_range      = "443"
+  source_address_prefix       = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: You should not allow 443 from the internet to your VM since the entrypoint is the AppGW. I would remove this rule.

File: exercise1/terraform/main.tf [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+  server_id = azurerm_mssql_server.ex1_sql_server.id
+
+  # prevent the possibility of accidental data loss
+  lifecycle {
+    prevent_destroy = false # for the sake of testing
+  }
+}
+
+#_______________________ Redis Database resources _______________________________________________
+resource "azurerm_redis_cache" "ex1_vm_redis" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: Missing private endpoint. See above.

File: exercise1/terraform/main.tf [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+}
+
+resource "azurerm_network_interface_security_group_association" "ex1_vm_netsecg_nic_asso" {
+  network_interface_id      = azurerm_network_interface.ex1_nic_vm.id
+  network_security_group_id = azurerm_network_security_group.ex1_vm_netsecg.id
+}
+
+#___________________________ VM associated resources ______________________________
+
+resource "azurerm_virtual_machine" "ex1_vm" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: Typically all config values should be variables but for the purposes of this exercise, hard coded is fine.

File: exercise1/terraform/main.tf [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+
+resource "azurerm_network_security_rule" "health_probe_inbound" {
+  name                        = "AllowHealthProbe"
+  priority                    = 1003
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "65200-65535"
+  source_address_prefix       = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: Huge security risk because you are allowing any source on ports "65200-65535" to reach your env. Limit it to the LoadBalancer tag as the source.

File: exercise1/terraform/main.tf [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+resource "azurerm_key_vault" "ex1_akv" {
+  name                      = "${var.rg_name}-akv-876987"
+  resource_group_name       = azurerm_resource_group.ex1.name
+  location                  = azurerm_resource_group.ex1.location
+  sku_name                  = "standard"
+  tenant_id                 = data.azurerm_client_config.current.tenant_id
+  enable_rbac_authorization = true
+}
+
+data "azurerm_role_definition" "keyvault_cert_user" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: Why did you add this resource? To validate the role exists?

File: exercise1/terraform/main.tf [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+  }
+  # # Bonus attempt
+  # waf_configuration {
+  #   enabled          = true
+  #   firewall_mode    = "Detection"
+  #   rule_set_version = 3.2
+  # }
+  trusted_root_certificate {
+    name                = local.cert_tls_ssl
+    key_vault_secret_id = azurerm_key_vault_certificate.ex1_cert_appgw.secret_id
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: This seems to be your SSL server certificate not the root CA. You need to pass the server cert instead.

File: exercise1/terraform/main.tf (hunk 6 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: I was looking into an implementation where the virtual machine extension calls the script from a file in the storage account then left this because i thought it was necessary for the `azurerm_mssql_database` which it seems that it is not. 

Removing

File: exercise1/terraform/main.tf (hunk 12 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: This is an artifact of a previous implementation but can now be removed as we can reference  builtin role definitions directly in the `azurerm_role_assignment` block.

Removing

File: exercise1/terraform/main.tf (hunk 11 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: I agree but the application gateway would not provision without these ports as it uses them for health probes, loopbacks, etc.

File: exercise1/terraform/main.tf (hunk 8 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: I put this here because I thought it was needed for the traffic from the application gateway but it seems that this is handled directly by the nic attached to the vm as well as the backend pool association?

Please confirm if the above is correct conceptually

Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: Removed.

File: exercise1/terraform/main.tf (hunk 7 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: good to know!

File: exercise1/terraform/main.tf (hunk 4 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: understood and removed

File: exercise1/terraform/main.tf (hunk 10 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: I prefer writing production quality code :)

I added vars for almost everything but ideally it would be one big object at the variables.tf level

File: exercise1/terraform/main.tf (hunk 5 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: i see. adjusting and making another one for redis

File: exercise1/terraform/main.tf (hunk 9 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: added a endpoint setup for redis

File: exercise1/terraform/main.tf (hunk 3 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: i matched exactly and added nothing infront and  also matched to the resource

File: exercise1/terraform/main.tf (hunk 2 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: i added rules defined in the documentation but it seems there exists https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mssql_firewall_rule

i am not sure if this is worth implementing or overkill.

File: exercise1/terraform/main.tf (hunk 13 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: Taking a stab at it and going to test with ssl_certificate.

I am unsure if you mean that the trusted root cert block is the wrong one to use or that i am passing the wrong cert from the akv

**Comment by Alex-Caesar** [comment]
Current known issues:
- Application Gateway and Backend Pool Probes do not work: seems to do with the NSG of the subnet for vm or the routing rules in the app gw
- I do not know how or where the cert is supposed to be. My best guess is the cert is for the app gw and is client facing. what is the desired implementation?

File: exercise1/terraform/main.tf (hunk 2 above)
Comment (jb-coalfire) [reply exercise1/terraform/main.tf]: no need.

File: exercise1/terraform/main.tf [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+# the Following two rules are ref from https://learn.microsoft.com/en-us/azure/azure-sql/managed-instance/connectivity-architecture-overview?view=azuresql&tabs=current#:~:text=You%20can%20use%20a%20network%20security%20group%20to%20control%20access%20to%20the%20SQL%20Managed%20Instance%20data%20endpoint%20by%20filtering%20traffic%20on%20port%201433%20and%20ports%2011000%2D11999%20when%20SQL%20Managed%20Instance%20is%20configured%20for%20redirect%20connections.
+resource "azurerm_network_security_rule" "filter1433" {
+  name                        = "filter1433"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Deny"
+  protocol                    = "*"
+  source_port_range           = "1433"
+  destination_port_range      = "1433"
+  source_address_prefix       = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: Why are you still allowing all access on 1433. Source should your VM subnet only.

File: exercise1/terraform/main.tf [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+  destination_address_prefix  = "*"
+  resource_group_name         = azurerm_resource_group.ex1.name
+  network_security_group_name = azurerm_network_security_group.ex1_sql_netsecg.name
+}
+
+resource "azurerm_network_security_rule" "filterRedirect" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: What is the purpose of this? You have a deny all after it.

File: exercise1/terraform/main.tf (hunk 8 above)
Comment (jb-coalfire) [reply exercise1/terraform/main.tf]: Traffic is handled through service tags, GatewayManager and AzureLoadBalancer i believe.

File: exercise1/terraform/main.tf (hunk 11 above)
Comment (jb-coalfire) [reply exercise1/terraform/main.tf]: use the service tag.
Allow incoming requests from the source as the GatewayManager service tag and Any destination.

File: exercise1/terraform/main.tf [hunk 16]
```diff
@@ -0,0 +1,537 @@
...
+  source_port_range           = "*"
+  destination_port_range      = "65200-65535"
+  source_address_prefix       = "*"
+  destination_address_prefix  = "*"
+  resource_group_name         = azurerm_resource_group.ex1.name
+  network_security_group_name = azurerm_network_security_group.ex1_app_gw_netsecg.name
+}
+
+# Deny all inbound traffic not explicitly allowed
+resource "azurerm_network_security_rule" "deny_all_inbound_app_gw" {
```
Comment (jb-coalfire) [inline exercise1/terraform/main.tf]: You need this allow before this deny:
AzureLoadBalancer	Any	Any	Any	Any	Allow

File: exercise1/terraform/main.tf (hunk 14 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: got it. now using service tags.

File: exercise1/terraform/main.tf (hunk 15 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: this is another necessary port range from the documentation. i noticed that i made a silly copy and paste error where the security rules were set to `Deny` instead of `Allow` adjusted now.

the other security rule was to ensure no other access as the last rule. this can be taken out if you would prefer

Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: actually took it out cause this is implied by the NSG and having it seems like bad practice.

File: exercise1/terraform/main.tf (hunk 11 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: got it

File: exercise1/terraform/main.tf (hunk 16 above)
Comment (Alex-Caesar) [reply exercise1/terraform/main.tf]: added `lb_inbound` if i understood correctly

File: exercise1/terraform/app_gw.tf [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+# Inbound Client traffic
+resource "azurerm_network_security_rule" "https_rule_app_gw" {
+  name                        = "AllowHTTPS"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "80"
```
Comment (jb-coalfire) [inline exercise1/terraform/app_gw.tf]: 80 is insecure and should not be used.

File: exercise1/terraform/app_gw.tf [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+
+resource "azurerm_network_security_rule" "health_probe_inbound" {
+  name                        = "AllowHealthProbe"
+  priority                    = 300
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "65200-65535"
+  source_address_prefix       = "*" #Terraform specific constraint requires internet traffic
```
Comment (jb-coalfire) [inline exercise1/terraform/app_gw.tf]: This needs to be corrected. You canot allow any source to access your environments on these ports.

File: exercise1/terraform/app_gw.tf [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+  destination_address_prefix  = "*"
+  resource_group_name         = azurerm_resource_group.ex1.name
+  network_security_group_name = azurerm_network_security_group.ex1_app_gw_netsecg.name
+}
+
+# Outbound rule
+resource "azurerm_network_security_rule" "outbound_internet" {
+  name                        = "outboundInternet"
+  priority                    = 400
+  direction                   = "Inbound"
```
Comment (jb-coalfire) [inline exercise1/terraform/app_gw.tf]: You are allowing any-any inbound. This is bad practice, bu i think you meant to make it outbound.

File: exercise1/terraform/databases_pe.tf [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+resource "azurerm_network_security_rule" "filter1433" {
+  name                        = "filter1433"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "*"
+  source_port_range           = "1433"
+  destination_port_range      = "1433"
+  source_address_prefix       = azurerm_subnet.ex1_subnet_vm.address_prefixes.0
+  destination_address_prefix  = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/databases_pe.tf]: Good practice is to set the destination instead of any. use the db subnet address_prefix instead.

File: exercise1/terraform/app_gw.tf [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+resource "azurerm_network_security_rule" "lb_inbound" {
+  name                        = "AllowLb"
+  priority                    = 200
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "*"
+  source_port_range           = "*"
+  destination_port_range      = "*"
+  source_address_prefix       = "AzureLoadBalancer"
+  destination_address_prefix  = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/app_gw.tf]: Good practice is to set the destination instead of any. use the Vm subnet address_prefix instead.

File: exercise1/terraform/databases_pe.tf [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+resource "azurerm_network_security_rule" "filterRedirect" {
+  name                        = "filterRedirect"
+  priority                    = 200
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "*"
+  source_port_range           = "11000-11999"
+  destination_port_range      = "11000-11999"
+  source_address_prefix       = azurerm_subnet.ex1_subnet_pe.address_prefixes.0
+  destination_address_prefix  = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/databases_pe.tf]: Same as above.

File: exercise1/terraform/variables.tf [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+variable "db_admin" {
+  type        = string
+  default     = "dbadmin"
+  description = "The username for the msSQL server"
+  sensitive   = true
+}
+
+variable "db_password" {
+  type        = string
+  default     = "Complex6DbPassw0rd!"
```
Comment (jb-coalfire) [inline exercise1/terraform/variables.tf]: bad practice to store password in code.

File: exercise1/terraform/variables.tf [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+variable "vm_admin" {
+  type        = string
+  default     = "vmadmin"
+  description = "The username for the vm"
+  sensitive   = true
+}
+
+variable "vm_password" {
+  type        = string
+  default     = "Complex6VmPassw0rd!"
```
Comment (jb-coalfire) [inline exercise1/terraform/variables.tf]: Same as above. Pass values as env vars instead or use tfvars file (with that file added in your .gitignore)

File: exercise1/terraform/vm.tf [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+resource "azurerm_network_security_rule" "https_rule_vm_gw" {
+  name                        = "AllowHTTPSGW"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "443"
+  source_address_prefix       = "GatewayManager"
+  destination_address_prefix  = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/vm.tf]: Good practice is to set the destination instead of any. use the VM subnet address_prefix instead.

File: exercise1/terraform/vm.tf [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+resource "azurerm_network_security_rule" "https_rule_vm_lb" {
+  name                        = "AllowHTTPSLB"
+  priority                    = 200
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "443"
+  source_address_prefix       = "AzureLoadBalancer"
+  destination_address_prefix  = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/vm.tf]: Same as above. Also since it is the same rule, i think you can add multiple source prefixes. Consolidate where possible for less management.

File: exercise1/terraform/vm.tf [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+  os_disk {
+    name                 = "${var.rg_name}_vm_os_disk"
+    caching              = var.vm_caching
+    storage_account_type = var.vm_storage_account_type
+  }
+
+  computer_name = var.vm_name
+
+  admin_username = var.vm_admin
+  admin_password = var.vm_password
```
Comment (jb-coalfire) [inline exercise1/terraform/vm.tf]: This should use ssh key instead of password. Better practice.

File: exercise1/terraform/app_gw.tf (hunk 17 above)
Comment (Alex-Caesar) [reply exercise1/terraform/app_gw.tf]: corrected. not sure why it was not 443.

File: exercise1/terraform/app_gw.tf (hunk 18 above)
Comment (Alex-Caesar) [reply exercise1/terraform/app_gw.tf]: i have set it to only allow from `GatewayManager`

File: exercise1/terraform/app_gw.tf (hunk 19 above)
Comment (Alex-Caesar) [reply exercise1/terraform/app_gw.tf]: yes, small mistake. need to pay attention to details.

File: exercise1/terraform/app_gw.tf (hunk 21 above)
Comment (Alex-Caesar) [reply exercise1/terraform/app_gw.tf]: done.

File: exercise1/terraform/databases_pe.tf (hunk 20 above)
Comment (Alex-Caesar) [reply exercise1/terraform/databases_pe.tf]: will do from now on!

File: exercise1/terraform/vm.tf (hunk 25 above)
Comment (Alex-Caesar) [reply exercise1/terraform/vm.tf]: gotcha

File: exercise1/terraform/vm.tf (hunk 26 above)
Comment (Alex-Caesar) [reply exercise1/terraform/vm.tf]: consolidated and good idea

File: exercise1/terraform/variables.tf (hunk 23 above)
Comment (Alex-Caesar) [reply exercise1/terraform/variables.tf]: no default now

File: exercise1/terraform/variables.tf (hunk 24 above)
Comment (Alex-Caesar) [reply exercise1/terraform/variables.tf]: no default now so pass at cli time

File: exercise1/terraform/vm.tf (hunk 27 above)
Comment (Alex-Caesar) [reply exercise1/terraform/vm.tf]: yes. commented out and have ssh key enabled now

File: exercise1/terraform/databases_pe.tf (hunk 22 above)
Comment (Alex-Caesar) [reply exercise1/terraform/databases_pe.tf]: got it.

File: exercise1/terraform/app_gw.tf [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+# Inbound Client traffic
+resource "azurerm_network_security_rule" "https_rule_app_gw" {
+  name                        = "AllowHTTPS"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "443"
+  source_address_prefixes     = [azurerm_subnet.ex1_subnet_app_gw.address_prefixes.0]
```
Comment (jb-coalfire) [inline exercise1/terraform/app_gw.tf]: This is backwards. You allow any since that is your entrypoint into your app. Destination is the GW subnet. They way you have it now you will not be able to access your app externally.

File: exercise1/terraform/app_gw.tf [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+resource "azurerm_network_security_rule" "health_probe_inbound" {
+  name                        = "AllowHealthProbe"
+  priority                    = 300
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "65200-65535"
+  source_address_prefix       = "GatewayManager"
+  destination_address_prefix  = "*"
```
Comment (jb-coalfire) [inline exercise1/terraform/app_gw.tf]: Should have destination prefix just like AllowLb rule. You can consolidate these 2 rules into one.

File: exercise1/terraform/variables.tf [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+variable "vm_admin" {
+  type        = string
+  default     = "vmadmin"
+  description = "The username for the vm"
+  sensitive   = true
+}
+
+variable "vm_password" {
```
Comment (jb-coalfire) [inline exercise1/terraform/variables.tf]: Not needed since you are using ssh key now.

File: exercise1/terraform/vm.tf [hunk 31]
```diff
@@ -0,0 +1,168 @@
...
+  os_disk {
+    name                 = "${var.rg_name}_vm_os_disk"
+    caching              = var.vm_caching
+    storage_account_type = var.vm_storage_account_type
+  }
+
+  computer_name = var.vm_name
+
+  admin_username = var.vm_admin
+  # admin_password = var.vm_password
```
Comment (jb-coalfire) [inline exercise1/terraform/vm.tf]: remove to avoid confusion.

File: exercise1/terraform/app_gw.tf (hunk 28 above)
Comment (Alex-Caesar) [reply exercise1/terraform/app_gw.tf]: yes, i have corrected it.

File: exercise1/terraform/variables.tf (hunk 30 above)
Comment (Alex-Caesar) [reply exercise1/terraform/variables.tf]: done.

File: exercise1/terraform/vm.tf (hunk 31 above)
Comment (Alex-Caesar) [reply exercise1/terraform/vm.tf]: done.

File: exercise1/terraform/app_gw.tf (hunk 29 above)
Comment (Alex-Caesar) [reply exercise1/terraform/app_gw.tf]: done. when i combined the two rules the destination port range was forced to become * since LB needs *.

File: exercise1/terraform/keyvault.tf [hunk 32]
```diff
@@ -0,0 +1,107 @@
+#________________________ AKV associated resources ______________________________________________
+resource "azurerm_key_vault" "ex1_akv" {
```
Comment (jb-coalfire) [inline exercise1/terraform/keyvault.tf:2]: AKV is public by default. [public_network_access_enabled](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault#public_network_access_enabled) - (Optional) Whether public network access is allowed for this Key Vault. Defaults to true. Best practice is to either set it to private or enable fw on it as a minimum.

Comment (Alex-Caesar) [reply exercise1/terraform/keyvault.tf:2]: good catch!
